Clarify objection period for lazy consensus in SKIP 1 (#7020)
scikit-image/scikit-image: doc/source/skips/1-governance.rst

:Author: Mark Harfouche
:Author: Lars Grüter
:Author: Egor Panfilov

> :Author: Marianne Corvellec

Author: Added you as an author as well @mkcor. :wink:

:Resolved: 2019-09-25
:Resolution: https://github.com/scikit-image/scikit-image/pull/4182
:skimage-Version: 0.16

> :Revised: 2023-06-12

Reviewer: Should we update to today's date?

  sentence (but no change of the scikit-image.org landing page or the “about”
  page), require approval by a core developer *and* no disagreement or requested
  changes by a core developer on the issue or pull request page (lazy

>   consensus). Generally, an objection period of one or two days should be given before
>   lazy consensus is assumed.

Reviewer: ```suggestion
  consensus). Core developers are expected to wait one or two days for others
  to give their opinion on the pull request if they’re not confident others
  would agree.
```

Reviewer 2: I like this suggestion; some of consensus operates on an understanding of what would cause conflict, and what would be fine to continue with.

Author: See https://github.com/scikit-image/scikit-image/pull/7020/commits/e036dec4204f31f69933c6803f3f3580dfae86c4. Languagetool recommend a comma after "confident". :thinking:

Reviewer: > Languagetool recommend a comma after "confident"

Hm, maybe because really it stands for "if they’re not confident *that* others would agree."

- **Code changes and major documentation changes** require agreement by *two*
  core developers *and* no disagreement or requested changes by a core developer

>   on the issue or pull-request page (lazy consensus). Generally, an objection period
>   of one week should be given before lazy consensus is assumed.

Reviewer: ```suggestion
  on the issue or pull request page (lazy consensus). In this case, core
  developers are expected to wait about a week for others to give their
  opinion on the pull request, if they’re not confident others would agree.
```

  sentence (but no change of the scikit-image.org landing page or the “about”
  page), require approval by a core developer *and* no disagreement or requested
  changes by a core developer on the issue or pull request page (lazy

>   consensus). Core developers are expected to wait one or two days for others to give

Reviewer: ```suggestion
  consensus). Core developers are expected to wait three or four days for others to give
```

>   their opinion if they’re not confident that others would agree.

Reviewer: Do the two instances of "others" above mean "other core developers?"

Author: Yes. Though I imagine that in practice and most cases any concerns by non-core devs would also be treated with the same politeness.

Reviewer: Right. Then let's not have it mean "other core developers." Thanks. The sentence is not the most beautiful but I'll settle. Maybe "voice" instead of "give?"

Author: > Then let's not have it mean "other core developers."

It's not clear to me, should I update it or are you happy with this?

Reviewer: I'm happy as is. If we meant "other core developers," I would have suggested:
```diff
-others
+other core developers
```
(to be explicit/unambiguous).

- **Code changes and major documentation changes** require agreement by *two*
  core developers *and* no disagreement or requested changes by a core developer
  on the issue or pull-request page (lazy consensus). In this case, core developers are

>   expected to wait about a week for others to give their opinion if they’re not

Reviewer: ```suggestion
  expected to wait about two weeks for others to give their opinion if they’re not
```

>   confident, others would agree.

Reviewer: ```suggestion
  confident, that others would agree.
```

>   on the issue or pull-request page (lazy consensus). In case of such changes, core

Reviewer: Not sure how to interpret "in case of such changes:"

* in case of such requests?
* in case the requested changes have been applied?

>   on the issue or pull-request page (lazy consensus). In case of such changes, core

Author: I think @soupault didn't feel that "in this case" was clear enough? I don't mind either version. To me both just reference the described changes above clearly enough.

>   on the issue or pull-request page (lazy consensus). In case of such changes, core

Author: I've never seen pull request written with a dash. If this is not a grammar thing I'd say "pull request" is the more consistent one with the docs. But feel welcome to apply the patch. :)

>   on the issue or pull-request page (lazy consensus). In case of such changes, core

Reviewer: > I've never seen pull request written with a dash. If this is not a grammar thing I'd say "pull request" is the more consistent one with the docs. But feel welcome to apply the patch. :)

My suggestion (above) is to replace "pull-request" with "pull request" precisely. Grammar-wise, it's definitely "a pull request" (noun); in "pull request page" (above) though, the noun is "page" and "pull request" acts as an adjective, that's why it wouldn't be left-field to hyphenate it (look up "compound adjectives"). Still, I suggest that we don't (in this case). :wink:

>   on the issue or pull-request page (lazy consensus). In case of such changes, core

Reviewer: > I think @soupault didn't feel that "in this case" was clear enough? I don't mind either version. To me both just reference the described changes above clearly enough.

Ok, thanks. Then I would say "in the case of disagreement or requested changes" to be clear; I agree with @soupault ("in this case" could mean "agreement by *two* core developers *and* no disagreement or requested changes (...)" since it's a single sentence/chunk right before "in this case").

>   requested changes, core developers are expected to wait about a week for others to

Reviewer 2: "about a week" is confusing; perhaps rather "at least N days", with N being 5?

Reviewer: I don't find it confusing personally, but I don't mind replacing it with something like "somewhere between 5 and 10 days." The problem I see with "at least N days" is that it provides only a lower bound. Then, the (implicit) upper bound becomes, in my understanding, by default, *6 months*, which is the time for the dormant bot to ping us back. I find it too long and unreasonable. I find it reasonable to wait "at least 5 days" (but not more than 10, 15, or 20). Let's make it clear again that this waiting period is understood *with no activity* -- if a fierce discussion is going on, obviously I'm not saying it should be interrupted after 10 days!

Reviewer: Ok.

- **Changes to the API principles** require a :ref:`skip` and follow the

>   decision-making process outlined above. Except in this case an objection period of
>   one month should be given.

Reviewer: ```suggestion
  decision-making process outlined above. But, in this case, the objection period
  should be one-month-long.
```

>   decision-making process outlined above. But, in this case, the objection period

Reviewer 2: A month seems very long. Do we need that much time?

Reviewer: Apparently, we do! It's been way over a month since three core devs have approved this PR, I'd love to merge it now, but I admit I'd prefer having @stefanv's approval as well.

Reviewer 2: OK to go ahead with this PR as-is, as to not hold it up, but I do suggest we pull it back to two weeks to give the dev team license to move more quickly.

>   should be one-month-long.

Reviewer: Ok, I'm fine with either time periods; @stefanv we can merge the PR with this change:
```suggestion
  should be two-week-long.
```

Author: That's a pretty substantial changes given that other's have already approved this. So I'd rather merge this as is for now and see what kind of experience we have with this. I imagine one-month is totally fine since I would start counting once the SPEC is going live / leaves draft mode.

Reviewer 2: OK, merging this for now; btw it should probably be "the objection period should be one month" or "one month long".

Reviewer 2: Actually, going to let you fix that and then merge :)

Reviewer: > That's a pretty substantial change given that other's have already approved this. So I'd rather merge this as is for now and see what kind of experience we have with this. I imagine one-month is totally fine since I would start counting once the SPEC is going live / leaves draft mode.

Ok.

Reviewer: > btw it should probably be "the objection period should be one month" or "one month long".

:thinking: I guess either work although they're different:

* "the period is one month" is definitely valid;
* "one-month-long" has adjective value as in "a one-month-long period" so you can say "the period is one-month-long" (right?).

Reviewer 2: I believe you can use "a one-month-long period" or "the period is one month long", but not "the period is one-month-long".
